### PR TITLE
encoding.utf8: replace uchar with rune

### PR DESCRIPTION
**Pull request by peppergrayxyz**
`ustring`  has been replaced with `[]rune` in V 0.3 https://github.com/vlang/v/discussions/14895

`uchar`  should therefore be replaced with `rune` in the `encoding.utf8` module:
- added  `rune`-method --><br />deprecated `uchar` -method

public: 
-  `get_rune(s string, index int) rune` --><br />`get_uchar(s string, index int) int`
- `is_rune_punct(r rune) bool` --><br />`is_uchar_punct(uchar int) bool`
- `is_rune_global_punct(r rune) bool` --><br />`is_uchar_global_punct(uchar int) bool`

private:
- `find_punct_in_table(in_code rune, in_table []rune) rune`  --><br />`find_punct_in_table(in_code int, in_table []int) int`
- `unicode_punct_western rune[]` --><br />`unicode_punct_western int[]`
- `unicode_punct rune[]` --><br />`unicode_punct int[]`

### CODE REVIEW

File: vlib/encoding/utf8/utf8_util.v
Comment (spytheman) [inline vlib/encoding/utf8/utf8_util.v:32]:
```suggestion
@[deprecated_after: '2024-11-17']
```

File: vlib/encoding/utf8/utf8_util.v
Comment (spytheman) [inline vlib/encoding/utf8/utf8_util.v:186]:
```suggestion
@[deprecated_after: '2024-11-17']
```

File: vlib/encoding/utf8/utf8_util.v
Comment (spytheman) [inline vlib/encoding/utf8/utf8_util.v:205]:
```suggestion
@[deprecated_after: '2024-11-17']
```

**Review by spytheman — APPROVED** [review]
Good work.

**Comment by peppergrayxyz** [comment]
The [docs](https://github.com/vlang/v/edit/master/doc/docs.md#L5701) are somewhat unclear regarding this point:
- What is a good buffer for a future date? Another 6 months?
- When should we use dates and when not? I found that most deprecated functions have a date, but the docs only say it can be used.

I would also create a PR to update the docs with this info (when a date should be used and that a future date should be used here).

**Comment by spytheman** [comment]
I usually use a date, that is close to 1 month ahead in the future, since it is easy to just bump the month number, without having to do date calculations.

**Comment by spytheman** [comment]
The goal of it is to avoid immediate CI breakage, for people with actively developed V projects, i.e. it is assumed that in the next 30 days, they would have compiled their projects manually at least once, and thus they would have seen the deprecation notices and fixed the uses.

**Comment by spytheman** [comment]
> Another 6 months?
A longer buffer like that may also work, for stuff that is harder to change, like the IError interface change or the ?!Type split, but for just renaming functions, it seems excessive to me.

**Comment by spytheman** [comment]
> When should we use dates and when not? I found that most deprecated functions have a date, but the docs only say it can be used.

Adding a date is always preferable in my opinion, for stuff in the vlib modules.

A change with a deprecation date, to me, it is like we are giving an immediate ultimatum to all V users after the merge. It should probably be done for functions that are found to be conceptually broken, and obsoleted by much better functionality, but other than that, I do not see a situation where it would not be just rude to users.

**Comment by spytheman** [comment]
As for documenting the current deprecation policy - that can be done, but perhaps as suggestions, not as rules. Other V projects with other maintainers, can reasonably choose another policy if they wish to 🤷🏻‍♂️ .

**Comment by spytheman** [comment]
There is a tool btw, that can help with finding old deprecations, that we run 1-2 times per year: `v run cmd/tools/show_ancient_deprecations.v 180` for example, if ran now, will show all deprecations before 180 days ago, i.e. before 2024-04-21 .

**Comment by peppergrayxyz** [comment]
Thanks a lot for taking the time and the great explanations! ☺️

**Pull request merged**
